pass snapshot through to urlForFindHasMany and urlForFindBelongsTo

trevorrjohn opened the pull request:
related to: https://github.com/emberjs/data/pull/4304

trevorrjohn commented:
Not sure what is going on with appveyor here. I took a look at the logs, but wasn't able to discern the problem.

pangratz commented on addon/adapters/rest.js:
```diff
@@ -1279,7 +1279,7 @@ if (isEnabled('ds-improved-ajax')) {
         case 'findHasMany':
         case 'findBelongsTo':
-          let url = this.buildURL(type.modelName, id, null, requestType);
+          let url = this.buildURL(type.modelName, id, snapshots, requestType);
```
We want to pass the `snapshot` here.

trevorrjohn replied:
thanks, I am still confused about when snapshot vs snapshots is passed in here. 

Is `snapshot` always a `DS.Snapshot` and this `snapshots` always `DS.SnapshotRecordArray`?

pangratz replied:
Yes. Basically the `snapshot` here represents the "owner" of the relationship.

pangratz commented on addon/-private/adapters/build-url-mixin.js:
```diff
@@ -165,7 +165,7 @@ export default Ember.Mixin.create({
    * @param {String} modelName
```
Can you add an entry here for the snapshot?

pangratz commented on addon/-private/adapters/build-url-mixin.js:
```diff
@@ -175,7 +175,7 @@ export default Ember.Mixin.create({
    * @param {String} modelName
```
Also here...

pangratz commented:
You also need to pass the `snapshot` as third argument [here](https://github.com/emberjs/data/blob/6cb550929caf1ab48eeff0947b1032b8fbe2a298/addon/adapters/rest.js#L690) and [here](https://github.com/emberjs/data/blob/6cb550929caf1ab48eeff0947b1032b8fbe2a298/addon/adapters/rest.js#L637). Since this hasn't been caught a test, can you modify tests for the [`findBelongsTo`](https://github.com/emberjs/data/blob/6cb550929caf1ab48eeff0947b1032b8fbe2a298/tests/integration/adapter/rest-adapter-test.js#L1848-L1881) and [`findHasMany`](https://github.com/emberjs/data/blob/6cb550929caf1ab48eeff0947b1032b8fbe2a298/tests/integration/adapter/rest-adapter-test.js#L1702-L1742). I guess an assertion that the passed `snapshot` is present or it is an instance of `DS.Snapshot` would be good.

Apart from my comments, this looks great!

trevorrjohn commented:
So I updated the docs and added the snapshot where you pointed to above. I think I am still a little lost on what is going on between the call to `findRecord` and the call to `findHasMany`. I did some digging, but wasn't really able to follow the code path of where `findRecord` starts loading the associations.

pangratz commented on addon/adapters/rest.js:
```diff
@@ -1154,6 +1154,10 @@ var RESTAdapter = Adapter.extend(BuildURLMixin, {
 
+    if (snapshot) {
+      query.adapterOptions = snapshot.adapterOptions
+    }
```
This isn't needed, the main thing this PR tries to solve is just passing the `snapshot` to the `urlForFindBelongsTo` and `urlForFindHasMany` hook. Sorry if my comments confused you here...

pangratz commented on tests/integration/adapter/rest-adapter-test.js:
```diff
@@ -1700,8 +1700,11 @@ test("findHasMany - returning an array populates the array", function(assert) {
   adapter.shouldBackgroundReloadRecord = () => false;
+  let optionsStub = { options: true };
   adapter.buildURL = function(type, id, snapshot, requestType) {
+    assert.equal(snapshot.adapterOptions, optionsStub);
```
An assertion that the `snapshot` is instance of `DS.Snapshot` is sufficient here. There is actually no way to pass `adapterOptions` here, so that's not needed.

pangratz commented on tests/integration/adapter/rest-adapter-test.js:
```diff
@@ -1726,7 +1729,7 @@ test("findHasMany - passes buildURL the requestType", function(assert) {
   });
 
-  run(store, 'findRecord', 'post', '1').then(assert.wait(function(post) {
+  run(store, 'findRecord', 'post', '1', { adapterOptions: optionsStub }).then(assert.wait(function(post) {
```
This can also be removed.

pangratz commented:
> I think I am still a little lost on what is going on between the call to findRecord and the call to findHasMany. I did some digging, but wasn't really able to follow the code path of where findRecord starts loading the associations.

The `urlForFindBelongsTo` and `urlForFindHasMany` hooks are not called as a result of a `store.findXXX` or something, but as the result of a linked relationship: it is possible that a relationship in `ember-data` is defined not via an `id` but via a `link` (for example defined [here](https://github.com/emberjs/data/blob/6cb550929caf1ab48eeff0947b1032b8fbe2a298/tests/integration/adapter/rest-adapter-test.js#L1867) in the test). The `urlForFindBelongsTo` hook is then called as a result of accessing the linked relationship [here](https://github.com/emberjs/data/blob/6cb550929caf1ab48eeff0947b1032b8fbe2a298/tests/integration/adapter/rest-adapter-test.js#L1877)...

trevorrjohn commented:
ok updated, thanks for walking me through that. I was lost there for a bit.

pangratz commented:
Yeah, this can be quite confusing. But it looks good now!

Thanks @trevorrjohn!
